File: 07-Healthcare-DevSecOps-Governance/services/data-store/routers/fhir.py
from typing import Annotated

from auth import Principal, require_roles
from dependencies import DatabaseSession, NormalizerDependency, StoreDependency
from fastapi import APIRouter, Depends, Query
from schemas import Page, SourceSummary, StoreEnvelope, StoreResult

router = APIRouter(tags=["FHIR"])
Writer = Annotated[Principal, Depends(require_roles("ingestion:writer", "platform_admin"))]
Reader = Annotated[
    Principal,
    Depends(require_roles("clinical:reader", "security", "compliance", "platform_admin")),
]


@router.post("/store/fhir", response_model=StoreResult, status_code=201)
async def store_fhir(
    body: StoreEnvelope,
    identity: Writer,
    session: DatabaseSession,
    normalize: NormalizerDependency,
    store: StoreDependency,
) -> StoreResult:
    return await store.insert(
        session, "fhir", normalize.normalize("fhir", identity.tenant_id, body)
    )


@router.get("/v1/dashboard/fhir/summary", response_model=SourceSummary)
async def fhir_summary(identity: Reader, session: DatabaseSession, store: StoreDependency):
    return await store.summary(session, "fhir", identity.tenant_id)


@router.get("/v1/dashboard/fhir/patients", response_model=Page)
async def fhir_patients(
    identity: Reader,
    session: DatabaseSession,
    store: StoreDependency,
    search: str = Query(default="", max_length=100),
    limit: int = Query(default=50, ge=1, le=200),
):
    items, total = await store.fhir_patients(session, identity.tenant_id, search, limit)
    return Page(items=items, total=total)

File: 07-Healthcare-DevSecOps-Governance/portal/lib/api/fhir.ts
import { api, unwrap } from "./client";
import type { FhirPatient, Page, SourceSummary } from "@/types/ingestion";

export const fhirApi = {
  summary: () => unwrap<SourceSummary>(api.get("/data-store/v1/dashboard/fhir/summary")),
  patients: (search = "") =>
    unwrap<Page<FhirPatient>>(api.get("/data-store/v1/dashboard/fhir/patients", { params: { search, limit: 50 } })),
  ingestBundle: (tenantId: string, bundle: object) =>
    unwrap<{ accepted: number; rejected: number }>(
      api.post("/ingestion/fhir/ingest", { tenant_id: tenantId, bundle }),
    ),
  streamResource: (tenantId: string, resourceType: string) =>
    unwrap<{ accepted: number; rejected: number }>(
      api.post("/ingestion/fhir/ingest", { tenant_id: tenantId, resource_type: resourceType }),
    ),
};
